src/slurm_plugin/jobinfomgtd.py

```diff
@@ -0,0 +1,184 @@
+        self.cpu_gpu_cost_ratio = config.getfloat(
+            "jobinfomgtd",
+            "cpu_gpu_cost_ratio",
+            fallback=self.DEFAULTS.get("cpu_gpu_cost_ratio"),
+        )
+        self.custom_discount_rate = config.getfloat(
+            "jobinfomgtd",
+            "custom_discount_rate",
+            fallback=self.DEFAULTS.get("custom_discount_rate"),
+        )
+
+        # Configure boto3 to retry 1 times by default
+        self._boto3_retry = config.getint("jobinfomgtd", "boto3_retry", fallback=self.DEFAULTS.get("max_retry"))
+        self._boto3_config = {"retries": {"max_attempts": self._boto3_retry, "mode": "standard"}}
+        # Configure proxy
+        proxy = config.get("jobinfomgtd", "proxy", fallback=self.DEFAULTS.get("proxy"))
+        if proxy != "NONE":
+            self._boto3_config["proxies"] = {"https": proxy}
+        self.boto3_config = Config(**self._boto3_config)
+        self.head_node_instance_id = config.get("jobinfomgtd", "instance_id", fallback="unknown")
+
+
+class JobInfoManager:
+    """The job info manager."""
+
+    def __init__(self, config):
+        """Initialize JobInfoManager."""
+        self._config = None
+        self._current_time = None
+        self.set_config(config)
+
+    def process_job_info(self):
+        """Get job information from scontrol and publish job information as json log events."""
+        jobs = get_jobs_info()
+        if jobs:
+            log.info("Find jobs information. Publishing job information to log...")
+        self._event_publisher.publish_job_events(jobs)
+
+    def set_config(self, config):
+        if self._config != config:
+            log.info("Applying new jobinfomgtd config: %s", config)
+            self._config = config
+            self._event_publisher = ClusterEventPublisher.create_with_default_publisher(
+                event_logger, config.cluster_name, "HeadNode", "jobinfomgtd", config.head_node_instance_id
+            )
+
+
+def _run_jobinfomgtd(config_file):
+    config = JobinfomgtdConfig(config_file)
+    job_info_manager = JobInfoManager(config=config)
+    while True:
+        # Get loop start time
+        start_time = datetime.now(tz=timezone.utc)
+        # Get program config
+        try:
+            config = JobinfomgtdConfig(config_file)
+            job_info_manager.set_config(config)
+        except Exception as e:
+            log.warning("Unable to reload daemon config from %s, using previous one.\nException: %s", config_file, e)
+        # Configure root logger
+        try:
+            fileConfig(config.logging_config, disable_existing_loggers=False)
+        except Exception as e:
+            log.warning(
+                "Unable to configure logging from %s, using default logging settings.\nException: %s",
+                config.logging_config,
+                e,
+            )
+        job_info_manager.process_job_info()
```

Comment:
Do we want to terminate if `process_job_info` raises an excepiton?

Reply:
no, we catch the exception in the main function, should that already handled this case?

Reply:
I thought `main` just re-raises the error?

Reply:
In the `main` function, the exception will be caught by the `except` block. The error message will be log out, and the exception will be re-raised. With the `@retry` decorator, after `LOOP_TIME` seconds, the `main` function will be invoked again, it doesn't terminate the entire process. But I agree it is good to add exception handling for process_job_info